src/Compilers/VisualBasic/Portable/Symbols/Source/SourceFieldSymbol.vb

```diff
@@ -167,11 +167,415 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+                    Next
+
+                    graph.Remove(field)
+
+                    CheckGraph(graph)
+
+                    ' Add the start of the cycle to the ordered list.
+                    order.Add(New FieldInfo(field, startsCycle:=True))
+
+                    lastUpdated = updated
+                End If
+
+                [set].Free()
+            End While
+
+            lastUpdated?.Free()
+            fieldsInvolvedInCycles?.Free()
+        End Sub
+
+        Private Shared Function GetStartOfFirstCycle(
+            graph As Dictionary(Of SourceFieldSymbol, DependencyInfo),
+            ByRef fieldsInvolvedInCycles As ArrayBuilder(Of SourceFieldSymbol)
+        ) As SourceFieldSymbol
+            Debug.Assert(graph.Count > 0)
+
+            If fieldsInvolvedInCycles Is Nothing Then
+                fieldsInvolvedInCycles = ArrayBuilder(Of SourceFieldSymbol).GetInstance(graph.Count)
+                ' We sort fields that belong to the same compilation by location to process cycles in deterministic order.
+                ' Relative order between compilations is not important, cycles do not cross compilation boundaries. 
+                fieldsInvolvedInCycles.AddRange(graph.Keys.GroupBy(Function(f) f.DeclaringCompilation).
+                    SelectMany(Function(g) g.OrderByDescending(Function(f1, f2) g.Key.CompareSourceLocations(f1.Locations(0), f2.Locations(0)))))
+            End If
+
+            Do
+                Dim field As SourceFieldSymbol = fieldsInvolvedInCycles.Pop()
+
+                If graph.ContainsKey(field) AndAlso IsPartOfCycle(graph, field) Then
+                    Return field
+                End If
+            Loop
+        End Function
+
+        Private Shared Function IsPartOfCycle(graph As Dictionary(Of SourceFieldSymbol, DependencyInfo), field As SourceFieldSymbol) As Boolean
+            Dim [set] = PooledHashSet(Of SourceFieldSymbol).GetInstance()
+            Dim stack = ArrayBuilder(Of SourceFieldSymbol).GetInstance()
+
+            Dim stopAt As SourceFieldSymbol = field
+            Dim result As Boolean = False
+            stack.Push(field)
+
+            While stack.Count > 0
+                field = stack.Pop()
+                Dim node = graph(field)
+
+                If node.Dependencies.Contains(stopAt) Then
+                    result = True
+                    Exit While
+                End If
+
+                For Each dependency In node.Dependencies
+                    If [set].Add(dependency) Then
+                        stack.Push(dependency)
+                    End If
+                Next
+            End While
+
+            stack.Free()
+            [set].Free()
+            Return result
+        End Function
+
+        ''' <summary>
+        ''' Should be overridden by types utilizing <see cref="GetConstantValueImpl(ConstantFieldsInProgress)"/> helper.
+        ''' </summary>
+        Protected Overridable Function GetLazyConstantTuple() As EvaluatedConstant
```

Comment:
Should we prefer abstract for these helpers?

Reply:
> Should we prefer abstract for these helpers?

I don't think so. These helpers are not on MethodSymbol, are used only by two derived types utilizing GetConstantValueImpl helper and aren't meant to be used for any other purpose. It is extremely unlikely that we would introduce another form of declaration for a constant field to the language. Even if that happens, the requirements are clearly documented and any test scenario would reveal that requirements aren't met. 

---
In reply to: [550684153](https://github.com/dotnet/roslyn/pull/50169#discussion_r550684153) [](ancestors = 550684153)